### PR TITLE
fix: load table from json

### LINKED ISSUE #1228: load_table_from_json interpolates string as int
#### Issue

using `load_table_from_json()` to load data into a string column fails when using a value that can be interpolated as an int (eg. "123")

see code snippet, which shows inconsistency with the other data loading apis

#### Environment details

  - OS type and version: macos 12.2.1
  - Python version: 3.9.5 (also errors on 3.8)
  - pip version: `pip --version`
  - `google-cloud-bigquery` version: 2.31.0 & 3.0.1

#### Code example

```python
import io
import json

from google.api_core.exceptions import BadRequest
from google.cloud.bigquery import Client, Table, SchemaField

data = [{"col_1": "123"}]

project = 'my-project'
bq = Client(project)

table = Table(f'{project}.aaa.data_load_test', schema=[SchemaField("col_1", "STRING"), ])
bq.delete_table(table, not_found_ok=True)
bq.create_table(table)


# WORKS: bq load from byte stream
with io.StringIO() as fp:
    json.dump(data, fp)
    fp.seek(0)
    bq.load_table_from_file(fp, table).result()


# WORKS: streaming api
bq.insert_rows_json(table, data)


try:
    # FAILS: bq load from python dict.
    # 123 is interpolated as an int, even though i supply it as a string
    bq.load_table_from_json(data, table).result()
except BadRequest as e:
    print("ERROR")
    print(e)


# WORKS: bq load from python dict, but 123d cannot be interpolated as an int
bq.load_table_from_json([{"col_1": "123d"}], table).result()

```

#### Stack trace

```
Traceback (most recent call last):
  File "/Users/ben.marengo/code/scratch/bq_load.py", line 28, in <module>
    bq.load_table_from_json(data, table).result()
  File "/Users/ben.marengo/code/pyenvs/scratch/lib/python3.9/site-packages/google/cloud/bigquery/job/base.py", line 728, in result
    return super(_AsyncJob, self).result(timeout=timeout, **kwargs)
  File "/Users/ben.marengo/code/pyenvs/scratch/lib/python3.9/site-packages/google/api_core/future/polling.py", line 137, in result
    raise self._exception
google.api_core.exceptions.BadRequest: 400 Provided Schema does not match Table my-project:aaa.data_load_test. Field col_1 has changed type from STRING to INTEGER
```

### PR DESCRIPTION
**Pull request by abecerrilsalas**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #1228 🦕

### CODE REVIEW

File: google/cloud/bigquery/client.py [hunk 1]
```diff
@@ -2757,7 +2760,13 @@ def load_table_from_json(
 
         job_config.source_format = job.SourceFormat.NEWLINE_DELIMITED_JSON
 
-        if job_config.schema is None:
+        # check whether the table already exists with a schema
+        # if it does already have a schema, we will not tell the `backend to use autodetect.
+        table = _table_arg_to_table(destination, default_project=self.project)
+        if table.schema:
+            job_config.autodetect = False
+            job_config.schema = table.schema
```
Comment (steffnay) [inline google/cloud/bigquery/client.py:2768]: Is it necessary to set `job_config.schema` here?

File: tests/system/test_client.py [hunk 2]
```diff
@@ -21,6 +21,7 @@
 import operator
 import os
 import pathlib
+import string
```
Comment (steffnay) [inline tests/system/test_client.py:24]: Is this used?

File: google/cloud/bigquery/client.py [hunk 3]
```diff
@@ -22,6 +22,7 @@
 import datetime
 import functools
 import gzip
+from http import client
```
Comment (steffnay) [inline google/cloud/bigquery/client.py:25]: Is this used?

File: google/cloud/bigquery/client.py [hunk 4]
```diff
@@ -44,6 +45,8 @@
 import uuid
 import warnings
 
+from numpy import true_divide
```
Comment (steffnay) [inline google/cloud/bigquery/client.py:48]: ditto re: unused import?

File: google/cloud/bigquery/client.py [hunk 5]
```diff
@@ -2757,7 +2759,12 @@ def load_table_from_json(
 
         job_config.source_format = job.SourceFormat.NEWLINE_DELIMITED_JSON
 
-        if job_config.schema is None:
+        # check whether the table already exists with a schema
+        # if it does already have a schema, we will not tell the `backend to use autodetect.
+        table = _table_arg_to_table(destination, default_project=self.project)
```
Comment (steffnay) [inline google/cloud/bigquery/client.py:2764]: This method doesn't make an API call to check whether the table exists, which is what we need to do. This method is just creating a local object representing the table. You want to reference the docs that demonstrate how to check whether a table exists in order to find the correct method to use.

File: tests/system/test_client.py [hunk 6]
```diff
@@ -859,6 +859,67 @@ def test_load_table_from_json_basic_use(self):
...
+        ]
+
+        dataset_id = _make_dataset_id("bq_system_test")
+        self.temp_dataset(dataset_id)
+        table_id = "{}.{}.load_table_from_json_bug_check_with_schema".format(
+            Config.CLIENT.project, dataset_id
+        )
+
+        # create an empty table with schema
+        table = helpers.retry_403(Config.CLIENT.create_table)(Table(table_id))
```
Comment (steffnay) [inline tests/system/test_client.py:879]: The test name and comment say this table should be created with a schema, but I don't see the schema here.

File: tests/system/test_client.py [hunk 7]
```diff
@@ -859,6 +859,67 @@ def test_load_table_from_json_basic_use(self):
...
+            json_rows, table_id, job_config=job_config
+        )
+        load_job.result()
+
+        table = Config.CLIENT.get_table(table)
+        # schema should be inferred from the json data
+        self.assertEqual(
+            tuple(table.schema),
+            (bigquery.SchemaField("age", "STRING", mode="NULLABLE"),),
+        )
```
Comment (steffnay) [inline tests/system/test_client.py:893]: I think we need to add an assert checking that the row values were correctly inserted as strings.

File: tests/system/test_client.py [hunk 8]
```diff
@@ -859,6 +859,67 @@ def test_load_table_from_json_basic_use(self):
...
+            Config.CLIENT.project, dataset_id
+        )
+
+        # Create the table before loading so that schema mismatch errors are
+        # identified.
+        table = helpers.retry_403(Config.CLIENT.create_table)(Table(table_id))
+        self.to_delete.insert(0, table)
+
+        job_config = bigquery.LoadJobConfig()
+        job_config.autodetect = True
```
Comment (steffnay) [inline tests/system/test_client.py:916]: I'm not following this portion. What is the` job_config` here doing?

File: tests/system/test_client.py [hunk 9]
```diff
@@ -859,6 +859,69 @@ def test_load_table_from_json_basic_use(self):
...
+        self.temp_dataset(dataset_id)
+        table_id = "{}.{}.load_table_from_json_bug_check_with_schema".format(
+            Config.CLIENT.project, dataset_id
+        )
+
+        # create an empty table with schema
+        table = helpers.retry_403(Config.CLIENT.create_table)(Table(table_id))
+        self.to_delete.insert(0, table)
+
+        job_config = bigquery.LoadJobConfig(schema=table_schema)
```
Comment (steffnay) [inline tests/system/test_client.py:883]: Do we need to provide schema here?

File: tests/system/test_client.py [hunk 10]
```diff
@@ -859,6 +859,69 @@ def test_load_table_from_json_basic_use(self):
...
+        )
+        load_job.result()
+
+        table = Config.CLIENT.get_table(table)
+        # schema should be inferred from the json data
+        # self.assertEqual(
+        #     tuple(table.schema),
+        #     (bigquery.SchemaField("age", "STRING", mode="NULLABLE"),),
+        # )
+        assert type(table.schema[0].field_type) is str
```
Comment (steffnay) [inline tests/system/test_client.py:895]: This is checking that the schema is as expected, but we want to assert that the values are correct.

File: tests/system/test_client.py [hunk 11]
```diff
@@ -859,6 +859,69 @@ def test_load_table_from_json_basic_use(self):
         self.assertEqual(tuple(table.schema), table_schema)
         self.assertEqual(table.num_rows, 2)
 
+    def test_load_table_from_json_bug_check_with_schema(self):
```
Comment (steffnay) [inline tests/system/test_client.py:862]:
```suggestion
    def test_load_table_from_json_table_exists(self):
```

File: tests/system/test_client.py [hunk 12]
```diff
@@ -859,6 +859,69 @@ def test_load_table_from_json_basic_use(self):
...
+
+        table = Config.CLIENT.get_table(table)
+        # schema should be inferred from the json data
+        # self.assertEqual(
+        #     tuple(table.schema),
+        #     (bigquery.SchemaField("age", "STRING", mode="NULLABLE"),),
+        # )
+        assert type(table.schema[0].field_type) is str
+
+    def test_load_table_from_json_bug_check_with_no_schema(self):
```
Comment (steffnay) [inline tests/system/test_client.py:897]:
```suggestion
    def test_load_table_from_json_bug_table_not_exists(self):
```

File: tests/system/test_client.py [hunk 13]
```diff
@@ -859,6 +859,69 @@ def test_load_table_from_json_basic_use(self):
...
+        ]
+        dataset_id = _make_dataset_id("bq_system_test")
+        self.temp_dataset(dataset_id)
+        table_id = "{}.{}.load_table_from_json_bug_check".format(
+            Config.CLIENT.project, dataset_id
+        )
+
+        # Create the table with no schema
+        table = helpers.retry_403(Config.CLIENT.create_table)(Table(table_id))
+        self.to_delete.insert(0, table)
```
Comment (steffnay) [inline tests/system/test_client.py:914]:
```suggestion
```

File: tests/system/test_client.py [hunk 14]
```diff
@@ -859,6 +859,69 @@ def test_load_table_from_json_basic_use(self):
...
+
+        # job_config = bigquery.LoadJobConfig()
+        job_config = bigquery.LoadJobConfig()
+        job_config.autodetect = True
+        load_job = Config.CLIENT.load_table_from_json(json_rows, table_id)
+        load_job.result()
+
+        table = Config.CLIENT.get_table(table)
+        self.assertTrue(job_config.autodetect)
```
Comment (steffnay) [inline tests/system/test_client.py:923]: Check that autodetect was used.

File: google/cloud/bigquery/client.py [hunk 15]
```diff
@@ -2757,7 +2760,14 @@ def load_table_from_json(
 
         job_config.source_format = job.SourceFormat.NEWLINE_DELIMITED_JSON
 
-        if job_config.schema is None:
+        # make table id
+        # table_id = "your-project.your_dataset.your_table"
+        destination = _table_arg_to_table_ref(destination, default_project=self.project)
```
Comment (chalmerlowe) [inline google/cloud/bigquery/client.py:2765]: Why are we calling `_table_arg_to_table_ref()` here?
The very first line of `get_table()` calls that function.

table_ref = _table_arg_to_table_ref(table, default_project=self.project)

**Comment by meredithslota** [comment]
Closing in favor of https://github.com/googleapis/python-bigquery/issues/1646.

**Pull request closed without merging**